kep/42-podgroup-coscheduling/README.md

```diff
@@ -0,0 +1,125 @@
+## Goals
+1. Base on the scheduling framework, implement the co-scheduling feature.
+2. Define a CRD name PodGroup to help pod scheduling.
+ 
+## Non-Goals
+Sort the job when we submit jobs to a cluster. Currently, we can only do this base on pods.
+ 
+## Use Cases
+Batch workloads such as Spark jobs, TensorFlow jobs that have to run altogether.
+
+## Design Details
+
+### PodGroup
+
+We define a CRD name PodGroup to help schedule, its definition is as follows:
+```build
+// PodGroupSpec represents the template of a pod group.
+type PodGroupSpec struct {
+	// MinMember defines the minimal number of members/tasks to run the pod group;
+	// if there's not enough resources to start all tasks, the scheduler
+	// will not start anyone.
+	MinMember uint32 `json:"minMember"`
+
+	// MaxScheduleTime defines the maximal time of members/tasks to wait before run the pod group;
+	MaxScheduleTime *metav1.Duration `json:"maxScheduleTime,omitempty"`
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Current phase of PodGroup.
+	Phase PodGroupPhase `json:"phase"`
+
+	// OccupiedBy marks the podgroup occupied by which group.
+	// Owner reference would be used to filled it, if not initialize, it is empty
+	OccupiedBy string `json:"occupiedBy,omitempty"`
+
+	// The number of actively running pods.
+	// +optional
+	Scheduled uint32 `json:"scheduled"`
+
+	// The number of actively running pods.
+	// +optional
+	Running uint32 `json:"running"`
+
+	// The number of pods which reached phase Succeeded.
+	// +optional
+	Succeeded uint32 `json:"succeeded"`
+
+	// The number of pods which reached phase Failed.
+	// +optional
+	Failed uint32 `json:"failed"`
+
+	// ScheduleStartTime of the group
+	ScheduleStartTime metav1.Time `json:"scheduleStartTime"`
+}
+```
+
+### Controller
+
+We define a controller to reconcile PodGroup status, and we can query the job status through describing the PodGroup. Onece a pod in a group failed, the Group Status is marked Failed. Controller would also help recover from abnormal cases, e.g. batch scheduling is interrupted due to
+cluster upgrade.
+
+### Extension points
+
+#### QueueSort
+
+To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:
+
+1. Sort based on Pod Priority
+2. Pod does not belong to any group
+3. PodGroup creation time
+4. Pod creation time
+
+#### PreFilter
+
+This extension helps pre-filter pods. It is useful, especially when there are not enough resources in a cluster. The overall flow works as below:
+
+1. Allow the pods that do not belong to any group.
+2. If there are no groups scheduling, we check resource, if enough, we allow the pod.
```

Comment:
Can you rephrase this? I suppose you want to say:

```suggestion
2. If there are no other pending pod groups - say all other pod groups have already been scheduled, we allow the pod when its resource requirement is satisfied.
```

I'm not quite sure what do you mean by `we check resource`, maybe you mean `MinResource`; if so, please reword it.

Reply:
This comment is still outstanding.